File: 2015/day3.js
const input = require('./input/day3')

const recipients = input => {
  const directions = input.split('')
  let x = 0, y = 0, houses = { '0,0': true }
  const move = {
    '^': () => y++,
    'v': () => y--,
    '<': () => x--,
    '>': () => x++
  }
  directions.forEach(direction => {
    move[direction]()
    houses[`${x},${y}`] = true
  })
  return Object.keys(houses).length
}
console.log(recipients(input))

const recipients2 = input => {
  const directions = input.split('')
  let x = 0, y = 0, roboX = 0, roboY = 0, houses = { '0,0': true }
  const move = {
    '^': bool => bool ? y++ : roboY++,
    'v': bool => bool ? y-- : roboY--,
    '<': bool => bool ? x-- : roboX--,
    '>': bool => bool ? x++ : roboX++
  }
  directions.forEach((direction, i)=> {
    const bool = i % 2 === 0
    move[direction](bool)
    bool ? houses[`${x},${y}`] = true : houses[`${roboX},${roboY}`] = true
  })
  return Object.keys(houses).length
}
console.log(recipients2(input))

File: 2015/day3.rb
input = File.read('./input/day3.txt').strip.split('')

def atLeastOnce(input)
  x, y = [0, 0]
  recipients = { { x: x, y: y } => true }
  input.each_with_object(recipients) do |move, recipients|
    case move
    when '<'
      x -= 1
    when '>'
      x += 1
    when '^' 
      y += 1
    when 'v'
      y -= 1
    end
    recipients[{x: x, y: y}] = true
  end
  return recipients
end

puts atLeastOnce(input).keys.length

santa, roboSanta = input.partition.each_with_index{|_, i| i.odd?}

puts atLeastOnce(santa).merge(atLeastOnce(roboSanta)).keys.length
